packages/algorand-js/src/interpreter/txn.ts

```diff
@@ -0,0 +1,307 @@
+import { Address, Transaction } from "algosdk";
+
+import { TealError } from "../errors/errors";
+import { ERRORS } from "../errors/errors-list";
+import { ZERO_ADDRESS } from "../lib/constants";
+import { toBytes } from "../lib/parse-data";
+import { StackElem, TxnField, TxnType } from "../types";
+import { Op } from "./opcode";
+
+// regular bytes
+const bytes = [
+  TxnField.ApprovalProgram,
+  TxnField.ClearStateProgram,
+  TxnField.ConfigAssetUnitName,
+  TxnField.ConfigAssetName,
+  TxnField.ConfigAssetURL,
+  TxnField.ConfigAssetMetadataHash
+];
+
+// 32 byte addresses
+const addr = [
+  TxnField.Receiver,
+  TxnField.CloseRemainderTo,
+  TxnField.VotePK,
+  TxnField.SelectionPK,
+  TxnField.AssetSender,
+  TxnField.AssetReceiver,
+  TxnField.AssetCloseTo,
+  TxnField.ConfigAssetManager,
+  TxnField.ConfigAssetReserve,
+  TxnField.ConfigAssetFreeze,
+  TxnField.ConfigAssetClawback,
+  TxnField.FreezeAssetAccount
+];
+
+// returns default value of txn field
+function defaultSpec (txField: TxnField): StackElem {
+  if (bytes.includes(txField)) {
+    return new Uint8Array(0);
+  }
+  if (addr.includes(txField)) {
+    return ZERO_ADDRESS;
+  }
+
+  return BigInt('0');
+}
+
+function assertBytesDefined (a?: Uint8Array): Uint8Array {
+  if (a === undefined) {
+    return new Uint8Array(0);
+  }
+  return a;
+}
+
+function assertUint64Defined (a?: number): bigint {
+  if (a === undefined) {
+    return BigInt('0');
+  }
+  return BigInt(a);
+}
+
+// return Zero Address if addr is undefined
+function assertAddr (addr?: Address): Uint8Array {
+  if (addr === undefined) {
+    return ZERO_ADDRESS;
+  }
+  return addr.publicKey;
+}
+
+/**
+ * Description: returns specific transaction field value from tx object
+ * @param txField: transaction field
+ * @param interpreter: interpreter
+ */
+export function txnSpecbyField (txField: TxnField, tx: Transaction, gtxs: Transaction[]): StackElem {
+  // common fields (for all transaction types)
+  switch (txField) {
```

Comment:
I have a suggestion, as we are moving logic to constructors, therefore a list of args will be passed to `txn` constructor.

Instead of having `enums` and `swtich` conditions. I suggest to have a map. This way we can make our code easily maintainable and remove repetitive code.
Ex:
```
// Suppose this is transaction interface
interface tran {
   a: number,
   b: string
}
// tx object
const tx: tran = {
  a: 3,
  b: "string"
}
// we can have a single map like this for txfield and similary for other fields
let mp: {[key: string]: string } = {
  "Fee": "a",
   "B": "b"
}

// To access these values, you can simply pass arg[0]
const val = mp[arg[0]];
res = tx[val as keyof tran];
// result is your value.
```
This way we remove too many swtich cases

Reply:
I have tested this code, this is working

Reply:
This has some loopholes. Continuing discussion about the structure on discord

Reply:
yes there maybe some cases, but it can be done